### PR TITLE
report degraded status if router certs are invalid

### PR DESCRIPTION
**Pull request by sanchezl**
Add a `RouterCertsDomainValidationController` that validates that a route's certificates can actually validate a route's hostname. 
This was done in response to https://bugzilla.redhat.com/show_bug.cgi?id=1744599.

### CODE REVIEW

File: pkg/operator2/routercerts/controller.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+)
+
+// RouterCertsDomainValidationController validates that router certs match the ingress domain
+type RouterCertsDomainValidationController struct {
+	operatorClient v1helpers.OperatorClient
+	cachesToSync   []cache.InformerSynced
+	queue          workqueue.RateLimitingInterface
+	eventRecorder  events.Recorder
+
+	ingressLister v12.IngressLister
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: bad import name.  configv1listers

File: pkg/operator2/routercerts/controller.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+	secretLister  v1.SecretNamespaceLister
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: corev1listers

File: pkg/operator2/routercerts/controller.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+	secretName    string
+	routerName    string
+}
+
+func NewRouterCertsDomainValidationController(
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder,
+	ingressLister v12.IngressLister,
+	secretLister v1.SecretNamespaceLister,
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: take a route lister, so you can get the actual route being used.

File: pkg/operator2/routercerts/controller.go [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+	secretName string,
+	routerName string,
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: route name

File: pkg/operator2/starter.go [hunk 5]
```diff
@@ -168,6 +170,16 @@ func RunOperator(ctx *controllercmd.ControllerContext) error {
 
 	configOverridesController := unsupportedconfigoverridescontroller.NewUnsupportedConfigOverridesController(operatorClient, ctx.EventRecorder)
 	logLevelController := loglevel.NewClusterOperatorLoggingController(operatorClient, ctx.EventRecorder)
+
+	routerCertsController := routercerts.NewRouterCertsDomainValidationController(
+		operatorClient,
+		ctx.EventRecorder,
+		configInformers.Config().V1().Ingresses().Lister(),
+		kubeInformersNamespaced.Core().V1().Secrets().Lister().Secrets(targetNamespace),
+		targetNamespace,
```
Comment (sanchezl) [inline pkg/operator2/starter.go]: I need the secret name here. (targetNamespace is wrong)

Comment (sanchezl) [reply pkg/operator2/starter.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 4 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 2 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 1 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+			Roots:         rootCerts,
+		})
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "InvalidRouterCert",
```
Comment (enj) [inline pkg/operator2/routercerts/controller.go]: `InvalidServerCertRouterCerts`

File: pkg/operator2/routercerts/controller.go [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	// verify certificate chain
+	for _, serverCert := range serverCerts {
+		_, err = serverCert.Verify(x509.VerifyOptions{
+			DNSName:       c.routeName + "." + ingress.Spec.Domain,
+			Intermediates: intermediateCerts,
+			Roots:         rootCerts,
+		})
+		if err == nil {
```
Comment (enj) [inline pkg/operator2/routercerts/controller.go]: not a fan of setting errs out of current scope and then checking on outer scope.  breaking this out into a function would make it easier to follow

File: pkg/operator2/routercerts/controller.go (hunk 3 above)
Comment (enj) [reply pkg/operator2/routercerts/controller.go]: Not sure what info we need to grab from the route (hostname?)

File: pkg/operator2/routercerts/controller.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+		cachesToSync:   nil,
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "RouterCertsDomainValidationController"),
+		eventRecorder:  eventRecorder,
+		ingressLister:  ingressLister,
+		secretLister:   secretLister,
+		secretName:     secretName,
+		routeName:      routerName,
+	}
+	operatorClient.Informer().AddEventHandler(controller.newEventHandler())
+	controller.cachesToSync = append(controller.cachesToSync, operatorClient.Informer().HasSynced)
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: and here

File: pkg/operator2/routercerts/controller.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+	if err != nil {
+		return err
+	}
+	if !management.IsOperatorManaged(spec.ManagementState) {
+		return nil
+	}
+
+	condition, err := c.validateRouterCertificates()
+	if err != nil {
+		return err
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/operator2/routercerts/controller.go]: > this should cause you to go degraded or something similar

No, it shouldn't.  Please don't do that.  That's how we've ended up with permafailing bugs.

File: pkg/operator2/routercerts/controller.go [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+	}
+
+	condition.Type = conditionRouterCertsDegradedType
+	_, _, err = v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(*condition))
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: this update happens in `validateRouterCertificates`

File: pkg/operator2/routercerts/controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/operator2/routercerts/controller.go]: > > this should cause you to go degraded or something similar
> 
> No, it shouldn't. Please don't do that. That's how we've ended up with permafailing bugs.

If you can shrink to a single set of lines like: 
```
	cond := newUpgradeableCondition(featureGates)
	if _, _, updateError := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(cond)); updateError != nil {
		return updateError
	}

```

I can buy it.

File: pkg/operator2/routercerts/controller.go [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+		secretName:     secretName,
+		routeName:      routerName,
+	}
+	operatorClient.Informer().AddEventHandler(controller.newEventHandler())
+	controller.cachesToSync = append(controller.cachesToSync, operatorClient.Informer().HasSynced)
+	return controller
+}
+
+func (c *RouterCertsDomainValidationController) sync() error {
+	spec, _, _, err := c.operatorClient.GetOperatorState()
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: if you want this method to set status, push this down into the function you're calling.

File: pkg/operator2/routercerts/controller.go [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+	ingress, err := c.ingressLister.Get("cluster")
+	if errors.IsNotFound(err) {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NoIngressConfig",
+			Message: fmt.Sprintf("ingresses.config.openshift.io/cluster: %v", err),
+		}, nil
+	}
+	if err != nil {
+		return nil, err
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: This doesn't make sense.  If you need to be degraded, be degraded.  I think all errors here are uniform.

File: pkg/operator2/routercerts/controller.go [hunk 13]
```diff
@@ -0,0 +1,251 @@
...
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "NoServerCertRouterCerts",
+			Message: fmt.Sprintf("secret/%v.spec.data[%v] no server certificates found", c.secretName, ingress.Spec.Domain),
+		}, nil
+	}
+
+	// verify certificate chain
+	for _, serverCert := range serverCerts {
+		_, err = serverCert.Verify(x509.VerifyOptions{
+			DNSName:       c.routeName + "." + ingress.Spec.Domain,
```
Comment (deads2k) [inline pkg/operator2/routercerts/controller.go]: @enj is this logic correct?

Comment (enj) [reply pkg/operator2/routercerts/controller.go]: Yes (route name and prefix are the same in the code base - `targetName`).

File: pkg/operator2/routercerts/controller.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 12 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 14]
```diff
@@ -0,0 +1,251 @@
...
+			Reason:  "NoRouterCertSecret",
+			Message: fmt.Sprintf("secret/%v: %v", c.secretName, err),
+		}, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// cert data should exist
+	data, ok := secret.Data[ingress.Spec.Domain]
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 15]
```diff
@@ -0,0 +1,251 @@
...
+	if !ok {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "MalformedRouterCertsPEM",
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 16]
```diff
@@ -0,0 +1,251 @@
...
+	certificates, err := crypto.CertsFromPEM(data)
+	if err != nil {
+		return &operatorv1.OperatorCondition{
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "MalformedRouterCertsPEM",
+			Message: fmt.Sprintf("secret/%v.spec.data[%v] certificates could not be parsed: %v", c.secretName, ingress.Spec.Domain, err),
+		}, nil
+	}
+
+	// categorize certificates
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 17]
```diff
@@ -0,0 +1,251 @@
...
+	var serverCerts []*x509.Certificate
+	intermediateCerts := x509.NewCertPool()
+	rootCerts, err := x509.SystemCertPool()
+	if err != nil {
+		klog.Infof("system cert pool not available: %v", err)
+		rootCerts = x509.NewCertPool()
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 18]
```diff
@@ -0,0 +1,251 @@
...
+	}
+	for _, certificate := range certificates {
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 7 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 6 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+		operatorClient: operatorClient,
+		cachesToSync:   nil,
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "RouterCertsDomainValidationController"),
+		eventRecorder:  eventRecorder,
+		ingressLister:  ingressLister,
+		secretLister:   secretLister,
+		secretName:     secretName,
+		routeName:      routerName,
+	}
+	operatorClient.Informer().AddEventHandler(controller.newEventHandler())
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 8 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+	// get ingress domain
+	ingressDomain := ingress.Spec.Domain
+	if len(ingressDomain) == 0 {
+		return newRouterCertsDegradedf("NoIngressDomain", "ingresses.config.openshift.io/cluster: no spec.domain specified")
+	}
+
+	// get router certs secret
+	secret, err := c.secretLister.Secrets(c.targetNamespace).Get(c.secretName)
+	if err != nil {
+		return newRouterCertsDegradedf("NoRouterCertSecret", "secret/%v could not be retrieved: %v", c.secretName, err)
```
Comment (enj) [inline pkg/operator2/routercerts/controller.go]: need namespace in the message

File: pkg/operator2/routercerts/controller.go [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+	// cert data should exist
+	data, _ := secret.Data[ingressDomain]
```
Comment (enj) [inline pkg/operator2/routercerts/controller.go]: nit: `data := secret.Data[ingressDomain]`

File: pkg/operator2/routercerts/controller.go [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+			klog.V(4).Infof("using CA %s as root", certificate.Subject.String())
+			verifyOptions.Roots.AddCert(certificate)
+		case certificate.IsCA:
+			klog.V(4).Infof("using CA %s as intermediate", certificate.Subject.String())
+			verifyOptions.Intermediates.AddCert(certificate)
+		default:
+			serverCerts = append(serverCerts, certificate)
+		}
+	}
+	if c.additionalRootCACertsPEM != nil {
```
Comment (enj) [inline pkg/operator2/routercerts/controller.go]: why do we need this

File: test/library/crypto.go [hunk 23]
```diff
@@ -0,0 +1,95 @@
+package library
```
Comment (enj) [inline test/library/crypto.go]: The lib-go crypto package should have helpers for most of this stuff.

File: test/e2e/certs_test.go [hunk 24]
```diff
@@ -21,24 +25,52 @@ func TestRouterCerts(t *testing.T) {
...
-	require.Equal(t, operatorManagedRouterCerts.Data, machineManagedRouterCerts.Data)
-	// there should be an entry for the ingress domain
-	ingress, err := configClient.ConfigV1().Ingresses().Get("cluster", metav1.GetOptions{})
+	routerCertsSecret, err := kubeClient.CoreV1().Secrets("openshift-config-managed").Get("router-certs", metav1.GetOptions{})
 	require.NoError(t, err)
-	require.Contains(t, operatorManagedRouterCerts.Data, ingress.Spec.Domain)
-	// the entry should contain certificates that validate the oauth service hostname
-	block, _ := pem.Decode(operatorManagedRouterCerts.Data[ingress.Spec.Domain])
-	cert, err := x509.ParseCertificate(block.Bytes)
+	// put bad certs in data
```
Comment (enj) [inline test/e2e/certs_test.go]: This seems to be racing against the ingress operator which will fix the cert at some point.  i.e. this test may be flakey.

File: pkg/operator2/routercerts/controller_test.go [hunk 25]
```diff
@@ -0,0 +1,365 @@
...
+	if der, err = x509.CreateCertificate(rand.Reader, template, signerCertificate, result.PrivateKey.Public(), signerPrivateKey); err != nil {
+		panic(err)
+	}
+	if result.Certificate, err = x509.ParseCertificate(der); err != nil {
+		panic(err)
+	}
+	return result
+}
+
+func newServerCertificate(signer *cryptoMaterials, hosts ...string) *cryptoMaterials {
```
Comment (enj) [inline pkg/operator2/routercerts/controller_test.go]: same comment about lib-go crypto

File: pkg/operator2/routercerts/controller_test.go [hunk 26]
```diff
@@ -0,0 +1,365 @@
...
+func withMissingRootCA(certificates []*cryptoMaterials) []*cryptoMaterials {
+	return certificates[1:]
+}
+
+type cryptoMaterials struct {
+	PrivateKey  *rsa.PrivateKey
+	Certificate *x509.Certificate
+}
+
+func newCertificateAuthorityCertificate(parent *cryptoMaterials) *cryptoMaterials {
```
Comment (enj) [inline pkg/operator2/routercerts/controller_test.go]: same comment about lib-go crypto

File: pkg/operator2/routercerts/controller_test.go [hunk 27]
```diff
@@ -0,0 +1,365 @@
...
+				),
+			)),
+			expectedStatus: operatorv1.ConditionTrue,
+			expectedReason: "NoServerCertRouterCerts",
+		},
+		{
+			name:           "InvalidServerCertRouterCerts",
+			ingress:        newIngress("cluster", withDomain("example.com")),
+			expectedStatus: operatorv1.ConditionTrue,
+			expectedReason: "InvalidServerCertRouterCerts",
```
Comment (enj) [inline pkg/operator2/routercerts/controller_test.go]: I would expect to see a few variations on this particular failure - like wrong DNS name, has roots but does not verify, etc.

File: pkg/operator2/route_test.go [hunk 28]
```diff
@@ -556,90 +556,6 @@ func Test_authOperator_handleRoute(t *testing.T) {
 				},
 			},
 		},
-		"route-secret-empty": {
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/route_test.go]: They are still used in some of the tests.

File: pkg/operator2/routercerts/controller.go [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	controllerWorkQueueKey           = "key"
+	conditionRouterCertsDegradedType = "RouterCertsDegraded"
+)
+
+// AdditionalCertsPEMFunc returns the bytes of a PEM encoded certificates
+type AdditionalCertsPEMFunc func() []byte
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: This used to be how I fed the proxy certs to the controller. The proxy certs are now injected into the system certs, so no need for this anymore.

File: pkg/operator2/routercerts/controller.go (hunk 22 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Removed

File: pkg/operator2/routercerts/controller.go [hunk 30]
```diff
@@ -0,0 +1,257 @@
...
+	certificates, err := crypto.CertsFromPEM(data)
+	if err != nil {
+		return newRouterCertsDegradedf("MalformedRouterCertsPEM", "secret/%v.spec.data[%v] certificates could not be parsed: %v", c.secretName, ingressDomain, err)
+	}
+
+	// categorize certificates
+	var serverCerts []*x509.Certificate
+	verifyOptions := x509.VerifyOptions{}
+	verifyOptions.Intermediates = x509.NewCertPool()
+	verifyOptions.Roots, err = c.systemCertPool()
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 21 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 20 above)
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 31]
```diff
@@ -0,0 +1,257 @@
...
+
+	condition := c.validateRouterCertificates()
+	if _, _, err = v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(*condition)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *RouterCertsDomainValidationController) validateRouterCertificates() *operatorv1.OperatorCondition {
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go [hunk 32]
```diff
@@ -0,0 +1,257 @@
...
+	eventRecorder  events.Recorder
+
+	ingressLister   configv1listers.IngressLister
+	secretLister    corev1listers.SecretLister
+	targetNamespace string
+	secretName      string
+	routeName       string
+
+	additionalRootCACertsPEM AdditionalCertsPEMFunc
+	systemCertPool           func() (*x509.CertPool, error)
```
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Done

File: pkg/operator2/routercerts/controller.go (hunk 29 above)
[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/routercerts/controller.go]: Removed

File: test/library/crypto.go (hunk 23 above)
Comment (sanchezl) [reply test/library/crypto.go]: lib-go crypto seems like overkill. These 2 functions (`NewServerCertificate` & `NewCertificateAuthorityCertificate`) do all I need for a test in memory and with minimal input. Eventually I'll move them to a library-go package for e2e tests.

File: test/library/library.go [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+// GenerateNameForTest generates a name of the form `prefix + test name + random string` that
+// can be used as a resource name. Convert the result to lowercase to use as a dns label.
+func GenerateNameForTest(t *testing.T, prefix string) string {
```
Comment (enj) [inline test/library/library.go]: Any reason not to just use the `GenerateName` field in object meta?

File: pkg/operator2/route.go [hunk 34]
```diff
@@ -146,89 +141,3 @@ func isIngressAdmitted(ingress routev1.RouteIngress) bool {
 func ingressToHost(ingress *configv1.Ingress) string {
 	return targetName + "." + ingress.Spec.Domain
 }
-
-func validateRouterSecret(routerSecret *corev1.Secret, ingress *configv1.Ingress) (string, error) {
-	// be careful not to print the routerSecret even when it is empty
-
-	if len(routerSecret.Data) == 0 {
```
[the remark this answers is not in the document]
Comment (enj) [reply pkg/operator2/route.go]: This seems to be unaddressed.

[the remark this answers is not in the document]
Comment (sanchezl) [reply pkg/operator2/route.go]: Seems to handle nil just fine. https://play.golang.org/p/Copb9NprxXL
Also, when I try it, does not panic.

File: test/library/library.go (hunk 33 above)
Comment (sanchezl) [reply test/library/library.go]: Done.

**Pull request merged**